File: app/src/main/java/ws/grigory/currencycalculator/calculator/Display.java
package ws.grigory.currencycalculator.calculator;

import static ws.grigory.currencycalculator.CCWidgetLarge.DF;

import java.util.ArrayList;

import ws.grigory.currencycalculator.Currency;

public class Display {
    public static final String EMPTY = "";
    public static final String ZERO = "0";
    public static final String ONE = "1";
    public ArrayList<Currency> currencies;
    public int currencyIndex;
    protected final StringBuilder display = new StringBuilder();

    public Display(int currencyIndex, ArrayList<Currency> currencies) {
        this.currencies = currencies;
        this.currencyIndex = currencyIndex;
    }

    public void shift() {
        if(currencyIndex < currencies.size()) {
            currencyIndex = currencyIndex == 0 ? currencies.size() - 1 : --currencyIndex;
            writeCurrencyValueToDisplay();
        }
    }
    public void writeCurrencyValueToDisplay() {
        clearDisplay();
        display.append((currencyIndex >= currencies.size() ||
                currencies.get(currencyIndex).value == 0) ? EMPTY :
                floatToString(currencies.get(currencyIndex).value));
    }

    public void calculateCurrencyValueToDisplay(float baseValue) {
        if(currencyIndex < currencies.size()) {
            currencies.get(currencyIndex).calculateValue(baseValue);
            writeCurrencyValueToDisplay();
        }
    }

    public void zeroingDisplay() {
        if(currencyIndex < currencies.size()){
            currencies.get(currencyIndex).value = 0;
        }
        display.setLength(0);
    }

    public void clearDisplay() {
        display.setLength(0);
    }

    protected String floatToString(float value) {
        if(value == Float.NEGATIVE_INFINITY || value == Float.POSITIVE_INFINITY) {
            return Float.toString(value);
        } else if(value == 0) {
            return ZERO;
        } else {
            return DF.format(Math.round(value * 100) / 100f);
        }
    }

    public String getValue() {
        return currencyIndex >= currencies.size() ? EMPTY : display.toString();
    }

    public String getCurrencyName() {
        return currencyIndex >= currencies.size() ? EMPTY : currencies.get(currencyIndex).name;
    }
}

File: app/src/main/java/ws/grigory/currencycalculator/calculator/Display.kt
package ws.grigory.currencycalculator.calculator

import ws.grigory.currencycalculator.Constants.DF
import ws.grigory.currencycalculator.Constants.EMPTY
import ws.grigory.currencycalculator.Constants.ZERO
import ws.grigory.currencycalculator.Currency
import kotlin.math.roundToLong

open class Display(var currencyIndex: Int, var currencies: ArrayList<Currency>) {
    protected open val display = StringBuilder()
    val value: String
        get() = if (currencyIndex < 0 || currencyIndex >= currencies.size) EMPTY else display.toString()
    val currencyName: String
        get() = if (currencyIndex < 0 || currencyIndex >= currencies.size) EMPTY else currencies[currencyIndex].name

    fun shift() {
        if (currencyIndex >= 0 && currencyIndex < currencies.size) {
            currencyIndex = if (currencyIndex == 0) currencies.size - 1 else --currencyIndex
            writeCurrencyValueToDisplay()
        }
    }

    fun calculateCurrencyValueToDisplay(baseValue: Float) {
        if (currencyIndex >= 0 && currencyIndex < currencies.size) {
            currencies[currencyIndex].calculateValue(baseValue)
            writeCurrencyValueToDisplay()
        }
    }

    open fun writeCurrencyValueToDisplay() {
        clearDisplay()
        display.append(
            if (currencyIndex >= currencies.size || currencies[currencyIndex].value == 0F) EMPTY
            else floatToString(currencies[currencyIndex].value)
        )
    }

    fun zeroingDisplay() {
        if (currencyIndex >= 0 && currencyIndex < currencies.size) {
            currencies[currencyIndex].value = 0F
        }
        clearDisplay()
    }

    fun clearDisplay() {
        display.apply { setLength(0) }
    }

    protected fun floatToString(value: Float): String {
        return when (value) {
            Float.NEGATIVE_INFINITY, Float.POSITIVE_INFINITY -> {
                value.toString()
            }

            0F -> {
                ZERO
            }

            else -> {
                DF.format((value * 100f).roundToLong() / 100f)
            }
        }
    }
}